File: grimdb/sdk/swift/Sources/GrimlockerSDK/GrimlockerClient.swift
import Foundation

/// Actor-based Grimlocker client — thread-safe by Swift's actor model.
///
/// ```swift
/// let client = GrimlockerClient(baseURL: URL(string: "http://127.0.0.1:36353")!,
///                               token: ProcessInfo.processInfo.environment["GRIMLOCKER_TOKEN"]!)
/// try await client.unlockVault(password: "master-password")
/// let passwords = try await client.listPasswords()
/// ```
public actor GrimlockerClient {
    private let baseURL: URL
    private let session: URLSession
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()

    public init(baseURL: URL, token: String, timeout: TimeInterval = 30) {
        self.baseURL = baseURL
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = timeout
        config.httpAdditionalHeaders = [
            "X-Grimlocker-Token": token,
            "Content-Type": "application/json",
        ]
        self.session = URLSession(configuration: config)
    }

    // ── Auth ──────────────────────────────────────────────────────────────────

    public func unlockVault(password: String) async throws {
        try await call("vault.unlock", payload: ["password": password])
    }

    public func lockVault() async throws {
        try await call("vault.logout", payload: [:] as [String: String])
    }

    public func vaultStatus() async throws -> VaultStatus {
        try await call("vault.status", payload: [:] as [String: String])
    }

    // ── Entries ───────────────────────────────────────────────────────────────

    public func listEntries(category: String? = nil) async throws -> [VaultEntry] {
        if let cat = category {
            return try await call("entry.query", payload: ["category": cat])
        }
        return try await call("entry.list", payload: [:] as [String: String])
    }

    public func getEntry(id: String) async throws -> VaultEntry {
        try await call("entry.read", payload: ["id": id])
    }

    public func createEntry(title: String, category: String, fields: [String: String]) async throws -> VaultEntry {
        try await call("entry.create", payload: ["title": title, "category": category, "fields": fields])
    }

    public func updateEntry(id: String, fields: [String: String]) async throws {
        try await call("entry.update", payload: ["id": id, "fields": fields])
    }

    public func deleteEntry(id: String) async throws {
        try await call("entry.delete", payload: ["id": id])
    }

    public func searchEntries(query: String, category: String? = nil) async throws -> [VaultEntry] {
        var p: [String: String] = ["query": query]
        if let c = category { p["category"] = c }
        return try await call("entry.search", payload: p)
    }

    // ── Typed helpers ─────────────────────────────────────────────────────────

    public func listPasswords() async throws -> [PasswordEntry] {
        try await listEntries(category: "PASSWORD").map(PasswordEntry.from)
    }

    public func createPassword(_ p: PasswordEntry) async throws -> String {
        try await createEntry(title: p.title, category: "PASSWORD", fields: p.fields).id
    }

    public func listSshKeys() async throws -> [SshKeyEntry] {
        try await listEntries(category: "SSH_KEY").map(SshKeyEntry.from)
    }

    public func createSshKey(_ k: SshKeyEntry) async throws -> String {
        try await createEntry(title: k.title, category: "SSH_KEY", fields: k.fields).id
    }

    public func listCertificates() async throws -> [CertificateEntry] {
        try await listEntries(category: "CERTIFICATE").map(CertificateEntry.from)
    }

    public func createCertificate(_ c: CertificateEntry) async throws -> String {
        try await createEntry(title: c.title, category: "CERTIFICATE", fields: c.fields).id
    }

    // ── File Vault ────────────────────────────────────────────────────────────

    public func listFolder(folderID: String = "") async throws -> FolderListing {
        try await call("file.list_folder", payload: ["folder_id": folderID])
    }

    public func createFolder(name: String, parentID: String = "") async throws -> FolderItem {
        try await call("file.create_folder", payload: ["name": name, "parent_id": parentID])
    }

    public func renameFolder(id: String, name: String) async throws {
        try await call("file.rename_folder", payload: ["id": id, "name": name])
    }

    public func deleteFolder(id: String) async throws {
        try await call("file.delete_folder", payload: ["id": id])
    }

    public func moveFile(manifestBlockID: String, folderID: String) async throws {
        try await call("file.move", payload: ["manifest_block_id": manifestBlockID, "folder_id": folderID])
    }

    public func uploadFile(
        _ data: Data, fileName: String,
        mimeType: String = "application/octet-stream",
        folderID: String = "",
        onProgress: ((UploadProgress) -> Void)? = nil
    ) async throws -> FileEntry {
        onProgress?(UploadProgress(bytesSent: 0, totalBytes: Int64(data.count)))
        let entry: FileEntry = try await call("file.ingest", payload: [
            "file_name": fileName, "mime_type": mimeType,
            "folder_id": folderID, "data_b64": data.base64EncodedString(),
        ])
        onProgress?(UploadProgress(bytesSent: Int64(data.count), totalBytes: Int64(data.count)))
        return entry
    }

    public func downloadFile(manifestBlockID: String) async throws -> Data {
        struct Resp: Decodable { let dataB64: String; enum CodingKeys: String, CodingKey { case dataB64 = "data_b64" } }
        let r: Resp = try await call("file.download", payload: ["manifest_block_id": manifestBlockID])
        guard let d = Data(base64Encoded: r.dataB64) else { throw GrimlockerError.decodingError("base64 decode failed") }
        return d
    }

    // ── Workspaces ────────────────────────────────────────────────────────────

    public func listWorkspaces() async throws -> [Workspace] {
        try await call("workspace.list", payload: [:] as [String: String])
    }

    public func createWorkspace(name: String) async throws -> Workspace {
        try await call("workspace.create", payload: ["name": name])
    }

    public func switchWorkspace(id: String) async throws {
        try await call("workspace.switch", payload: ["id": id])
    }

    public func renameWorkspace(id: String, name: String) async throws {
        try await call("workspace.rename", payload: ["id": id, "name": name])
    }

    public func deleteWorkspace(id: String) async throws {
        try await call("workspace.delete", payload: ["id": id])
    }

    // ── Sync ──────────────────────────────────────────────────────────────────

    public func listSyncPeers() async throws -> SyncStatus {
        try await call("sync.list_peers", payload: [:] as [String: String])
    }

    public func triggerSync() async throws {
        try await call("sync.trigger", payload: [:] as [String: String])
    }

    // ── Audit ─────────────────────────────────────────────────────────────────

    public func listAuditEvents(n: Int = 50) async throws -> [AuditEvent] {
        try await call("audit.list", payload: ["n": n])
    }

    // ── Health ────────────────────────────────────────────────────────────────

    public func healthCheck() async throws -> VaultStatus { try await vaultStatus() }

    // ── Internal ─────────────────────────────────────────────────────────────

    @discardableResult
    private func call<Payload: Encodable, Response: Decodable>(
        _ action: String, payload: Payload
    ) async throws -> Response {
        var req = URLRequest(url: baseURL.appendingPathComponent("api/v1"))
        req.httpMethod = "POST"
        req.httpBody = try encoder.encode(["action": AnyEncodable(action), "payload": AnyEncodable(payload)])

        let (data, resp) = try await session.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw GrimlockerError.network("non-HTTP response") }

        if http.statusCode < 200 || http.statusCode >= 300 {
            if let err = try? decoder.decode(APIError.self, from: data) {
                throw GrimlockerError.daemon(code: err.errorCode ?? 0, message: err.error ?? "HTTP \(http.statusCode)")
            }
            throw GrimlockerError.network("HTTP \(http.statusCode)")
        }
        return try decoder.decode(Response.self, from: data)
    }

    @discardableResult
    private func call<Payload: Encodable>(_ action: String, payload: Payload) async throws -> Void {
        let _: EmptyResponse = try await call(action, payload: payload)
    }

    private struct APIError: Decodable {
        let error: String?
        let errorCode: Int?
        enum CodingKeys: String, CodingKey { case error; case errorCode = "error_code" }
    }
    private struct EmptyResponse: Decodable {}
}

// MARK: - Error

public enum GrimlockerError: Error, LocalizedError {
    case network(String)
    case daemon(code: Int, message: String)
    case decodingError(String)

    public var errorDescription: String? {
        switch self {
        case .network(let m):            return "Network error: \(m)"
        case .daemon(let c, let m):      return "Daemon error (\(c)): \(m)"
        case .decodingError(let m):      return "Decoding error: \(m)"
        }
    }
}

// Type-erased Encodable for mixed-type dictionaries
private struct AnyEncodable: Encodable {
    let value: Any
    init(_ v: Any) { value = v }
    func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()
        switch value {
        case let v as String:  try c.encode(v)
        case let v as Int:     try c.encode(v)
        case let v as Bool:    try c.encode(v)
        case let v as [String: String]: try c.encode(v)
        case let v as [String: Int]:    try c.encode(v)
        case let v as Encodable:
            let wrapper = EncodableWrapper(v)
            try wrapper.encode(to: encoder)
        default: try c.encodeNil()
        }
    }
}
private struct EncodableWrapper: Encodable {
    let base: Encodable
    init(_ base: Encodable) { self.base = base }
    func encode(to encoder: Encoder) throws { try base.encode(to: encoder) }
}

File: grimdb/sdk/csharp/src/GrimlockerClient.cs
using System.Net.Http.Json;
using System.Text.Json;
using Grimlocker.Models;

namespace Grimlocker;

/// <summary>
/// Async HTTP client for the Grimlocker vault daemon.
/// Uses the /api/v1 JSON endpoint — no external dependencies.
/// </summary>
/// <example>
/// <code>
/// await using var client = new GrimlockerClient("http://127.0.0.1:36353", token);
/// await client.UnlockVaultAsync("master-password");
/// var passwords = await client.ListPasswordsAsync();
/// </code>
/// </example>
public class GrimlockerClient : IAsyncDisposable
{
    private readonly HttpClient   _http;
    private readonly string       _baseUrl;
    private static readonly JsonSerializerOptions _json = new()
        { PropertyNameCaseInsensitive = true };

    private readonly object _circuitLock = new();
    private int _consecutiveFailures = 0;
    private DateTimeOffset _circuitOpenUntil = DateTimeOffset.MinValue;

    /// <param name="baseUrl">Daemon base URL, e.g. "http://127.0.0.1:36353"</param>
    /// <param name="token">GRIMLOCKER_TOKEN from daemon stdout</param>
    public GrimlockerClient(string baseUrl, string token)
    {
        _baseUrl = baseUrl.TrimEnd('/');
        _http = new HttpClient();
        _http.DefaultRequestHeaders.Add("X-Grimlocker-Token", token);
    }

    // ── Auth ──────────────────────────────────────────────────────────────────

    public Task UnlockVaultAsync(string password, CancellationToken ct = default)
        => CallAsync("vault.unlock", new { password }, ct);

    public Task LockVaultAsync(CancellationToken ct = default)
        => CallAsync("vault.logout", new { }, ct);

    public async Task<VaultStatus> VaultStatusAsync(CancellationToken ct = default)
        => await CallAsync<VaultStatus>("vault.status", new { }, ct);

    // ── Entries ───────────────────────────────────────────────────────────────

    public async Task<IReadOnlyList<VaultEntry>> ListEntriesAsync(string? category = null, CancellationToken ct = default)
    {
        var payload = category is null ? new { } as object : new { category };
        var action  = category is null ? "entry.list" : "entry.query";
        return await CallAsync<List<VaultEntry>>(action, payload, ct) ?? [];
    }

    public async Task<VaultEntry> GetEntryAsync(string id, CancellationToken ct = default)
        => await CallAsync<VaultEntry>("entry.read", new { id }, ct)
           ?? throw new GrimlockerException($"Entry not found: {id}", -10);

    public async Task<VaultEntry> CreateEntryAsync(string title, string category,
        Dictionary<string, string> fields, CancellationToken ct = default)
        => await CallAsync<VaultEntry>("entry.create", new { title, category, fields }, ct)
           ?? throw new GrimlockerException("Create returned no entry", -30);

    public Task UpdateEntryAsync(string id, Dictionary<string, string> fields, CancellationToken ct = default)
        => CallAsync("entry.update", new { id, fields }, ct);

    public Task DeleteEntryAsync(string id, CancellationToken ct = default)
        => CallAsync("entry.delete", new { id }, ct);

    public async Task<IReadOnlyList<string>> CreateEntriesBatchAsync(IReadOnlyList<BatchEntryInput> items, CancellationToken ct = default)
    {
        var ids = new List<string>(items.Count);
        foreach (var item in items)
        {
            var entry = await CreateEntryAsync(item.Title, item.Category, item.Fields, ct);
            ids.Add(entry.Id);
        }
        return ids;
    }

    public async Task DeleteEntriesBatchAsync(IReadOnlyList<string> ids, CancellationToken ct = default)
    {
        foreach (var id in ids)
        {
            await DeleteEntryAsync(id, ct);
        }
    }

    public async Task<IReadOnlyList<VaultEntry>> SearchEntriesAsync(string query, string? category = null, CancellationToken ct = default)
        => await CallAsync<List<VaultEntry>>("entry.search", new { query, category }, ct) ?? [];

    // ── Typed helpers ─────────────────────────────────────────────────────────

    public async Task<IReadOnlyList<PasswordEntry>> ListPasswordsAsync(CancellationToken ct = default)
    {
        var entries = await ListEntriesAsync("PASSWORD", ct);
        return entries.Select(PasswordEntry.FromEntry).ToList();
    }

    public async Task<string> CreatePasswordAsync(PasswordEntry p, CancellationToken ct = default)
    {
        var entry = await CreateEntryAsync(p.Title, "PASSWORD", p.ToFields(), ct);
        return entry.Id;
    }

    public async Task<IReadOnlyList<SshKeyEntry>> ListSshKeysAsync(CancellationToken ct = default)
    {
        var entries = await ListEntriesAsync("SSH_KEY", ct);
        return entries.Select(SshKeyEntry.FromEntry).ToList();
    }

    public async Task<string> CreateSshKeyAsync(SshKeyEntry k, CancellationToken ct = default)
    {
        var entry = await CreateEntryAsync(k.Title, "SSH_KEY", k.ToFields(), ct);
        return entry.Id;
    }

    public async Task<IReadOnlyList<CertificateEntry>> ListCertificatesAsync(CancellationToken ct = default)
    {
        var entries = await ListEntriesAsync("CERTIFICATE", ct);
        return entries.Select(CertificateEntry.FromEntry).ToList();
    }

    public async Task<string> CreateCertificateAsync(CertificateEntry c, CancellationToken ct = default)
    {
        var entry = await CreateEntryAsync(c.Title, "CERTIFICATE", c.ToFields(), ct);
        return entry.Id;
    }

    // ── File Vault ────────────────────────────────────────────────────────────

    public async Task<FolderListing> ListFolderAsync(string? folderId = null, CancellationToken ct = default)
        => await CallAsync<FolderListing>("file.list_folder", new { folder_id = folderId ?? "" }, ct)
           ?? new FolderListing { Folders = [], Files = [] };

    public async Task<FolderItem> CreateFolderAsync(string name, string? parentId = null, CancellationToken ct = default)
        => await CallAsync<FolderItem>("file.create_folder", new { name, parent_id = parentId ?? "" }, ct)
           ?? throw new GrimlockerException("Folder creation failed");

    public Task RenameFolderAsync(string id, string name, CancellationToken ct = default)
        => CallAsync("file.rename_folder", new { id, name }, ct);

    public Task DeleteFolderAsync(string id, CancellationToken ct = default)
        => CallAsync("file.delete_folder", new { id }, ct);

    public Task MoveFileAsync(string manifestBlockId, string folderId, CancellationToken ct = default)
        => CallAsync("file.move", new { manifest_block_id = manifestBlockId, folder_id = folderId }, ct);

    public async Task<FileEntry> UploadFileAsync(
        byte[] data, string fileName, string mimeType,
        string? folderId = null,
        IProgress<(long sent, long total)>? progress = null,
        CancellationToken ct = default)
    {
        progress?.Report((0, data.Length));
        var dataB64 = Convert.ToBase64String(data);
        var entry = await CallAsync<FileEntry>("file.ingest", new
        {
            file_name = fileName, mime_type = mimeType,
            folder_id = folderId ?? "", data_b64 = dataB64
        }, ct) ?? throw new GrimlockerException("Upload returned no entry");
        progress?.Report((data.Length, data.Length));
        return entry;
    }

    public async Task<byte[]> DownloadFileAsync(string manifestBlockId, CancellationToken ct = default)
    {
        var result = await CallAsync<DownloadResult>("file.download",
            new { manifest_block_id = manifestBlockId }, ct);
        if (result?.DataB64 is null) throw new GrimlockerException("Download returned no data");
        return Convert.FromBase64String(result.DataB64);
    }

    // ── Workspaces ────────────────────────────────────────────────────────────

    public async Task<IReadOnlyList<Workspace>> ListWorkspacesAsync(CancellationToken ct = default)
        => await CallAsync<List<Workspace>>("workspace.list", new { }, ct) ?? [];

    public async Task<Workspace> CreateWorkspaceAsync(string name, CancellationToken ct = default)
        => await CallAsync<Workspace>("workspace.create", new { name }, ct)
           ?? throw new GrimlockerException("Workspace creation failed");

    public Task SwitchWorkspaceAsync(string id, CancellationToken ct = default)
        => CallAsync("workspace.switch", new { id }, ct);

    public Task RenameWorkspaceAsync(string id, string name, CancellationToken ct = default)
        => CallAsync("workspace.rename", new { id, name }, ct);

    public Task DeleteWorkspaceAsync(string id, CancellationToken ct = default)
        => CallAsync("workspace.delete", new { id }, ct);

    // ── Sync ──────────────────────────────────────────────────────────────────

    public async Task<SyncStatus> ListSyncPeersAsync(CancellationToken ct = default)
        => await CallAsync<SyncStatus>("sync.list_peers", new { }, ct)
           ?? new SyncStatus([], 0, "");

    public Task TriggerSyncAsync(CancellationToken ct = default)
        => CallAsync("sync.trigger", new { }, ct);

    // ── Audit ─────────────────────────────────────────────────────────────────

    public async Task<IReadOnlyList<AuditEvent>> ListAuditEventsAsync(int n = 50, CancellationToken ct = default)
        => await CallAsync<List<AuditEvent>>("audit.list", new { n }, ct) ?? [];

    // ── Tools ─────────────────────────────────────────────────────────────────

    public async Task<SshKeyResult> GenerateSshKeyAsync(string comment = "", bool saveToVault = true, CancellationToken ct = default)
        => await CallAsync<SshKeyResult>("tool.ssh_keygen", new { comment, save_to_vault = saveToVault }, ct)
           ?? throw new GrimlockerException("SSH key generation failed");

    public async Task<string> GetRecoveryPhraseAsync(string password, CancellationToken ct = default)
    {
        var result = await CallAsync<RecoveryResult>("vault.recovery_phrase", new { password }, ct);
        return result?.Phrase ?? throw new GrimlockerException("Recovery phrase retrieval failed");
    }

    // ── Health ────────────────────────────────────────────────────────────────

    public async Task<VaultStatus> HealthCheckAsync(CancellationToken ct = default)
        => await VaultStatusAsync(ct);

    // ── Internal ─────────────────────────────────────────────────────────────

    private async Task CallAsync(string action, object payload, CancellationToken ct)
    {
        await ExecuteWithRetryAsync(action, payload, ct);
    }

    private async Task<T?> CallAsync<T>(string action, object payload, CancellationToken ct)
    {
        var body = await ExecuteWithRetryAsync(action, payload, ct);
        return JsonSerializer.Deserialize<T>(body, _json);
    }

    private async Task<string> ExecuteWithRetryAsync(string action, object payload, CancellationToken ct)
    {
        lock (_circuitLock)
        {
            if (DateTimeOffset.UtcNow < _circuitOpenUntil)
                throw new CircuitBreakerOpenException("circuit breaker is open");
        }

        for (int attempt = 0; attempt < 4; attempt++)
        {
            string body = "";
            int statusCode = 0;
            bool isNetworkError = false;

            try
            {
                using var resp = await _http.PostAsJsonAsync($"{_baseUrl}/api/v1",
                    new { action, payload }, _json, ct);
                statusCode = (int)resp.StatusCode;
                body = await resp.Content.ReadAsStringAsync(ct);
            }
            catch (OperationCanceledException)
            {
                throw;
            }
            catch (HttpRequestException)
            {
                isNetworkError = true;
            }

            if (!isNetworkError && statusCode >= 200 && statusCode < 300)
            {
                lock (_circuitLock) { _consecutiveFailures = 0; }
                return body;
            }

            bool isClientError = !isNetworkError && statusCode >= 400 && statusCode < 500;

            if (isClientError || attempt == 3)
            {
                lock (_circuitLock)
                {
                    _consecutiveFailures++;
                    if (_consecutiveFailures >= 5)
                        _circuitOpenUntil = DateTimeOffset.UtcNow.AddSeconds(30);
                }

                if (isNetworkError)
                    throw new GrimlockerException($"network error ({action})", -104);

                int code = 0;
                string msg = body;
                try
                {
                    using var doc = JsonDocument.Parse(body);
                    code = doc.RootElement.TryGetProperty("error_code", out var c) ? c.GetInt32() : 0;
                    msg = doc.RootElement.TryGetProperty("error", out var e) ? e.GetString() ?? body : body;
                }
                catch { /* ignore parse errors */ }
                throw new GrimlockerException($"{GrimlockerException.NameOf(code)} ({action}): {msg}", code);
            }

            int delayMs = Math.Min(2000, 100 * (1 << attempt));
            await Task.Delay(delayMs, ct);
        }

        throw new GrimlockerException("retry exhaustion", -104);
    }

    public async ValueTask DisposeAsync() => _http.Dispose();

    private record DownloadResult([property: System.Text.Json.Serialization.JsonPropertyName("data_b64")] string? DataB64);
    private record RecoveryResult([property: System.Text.Json.Serialization.JsonPropertyName("phrase")] string? Phrase);
}

public class CircuitBreakerOpenException : GrimlockerException
{
    public CircuitBreakerOpenException(string message) : base(message, -200) { }
}

public sealed record BatchEntryInput(string Title, string Category, Dictionary<string, string> Fields);

public sealed record SshKeyResult
{
    [System.Text.Json.Serialization.JsonPropertyName("public_key")]
    public string PublicKey { get; init; } = "";
    [System.Text.Json.Serialization.JsonPropertyName("fingerprint")]
    public string Fingerprint { get; init; } = "";
    [System.Text.Json.Serialization.JsonPropertyName("entry_id")]
    public string? EntryId { get; init; }
}
